src/str.rs

```diff
@@ -491,7 +491,7 @@ mod test {
     let b = "ababcd";
 
     fn f(i: &str) -> IResult<&str, &str> {
-      recognize(many(1.., alt((tag("a"), tag("b")))))(i)
+      recognize::<_, Vec<&str>, _, _>(many(1.., alt((tag("a"), tag("b")))))(i)
```

Comment:
I've been going back and forth on this idea partly because of type inference issues when `fold_many*` easily let's people do the same thing.  I was surprised; they came up less than I expected.

> This will likely remove the optimizatio of preallocating vector capacity, but we can get this back with a wrapper type if needed, and being able to created directly a hashmap or other collections instead of going through a vec first will be way more useful

`fold_many` is likely lighter weight than a newtype (code wise).  Maybe we should provide a `fold_many` example that sets the capacity and have `many` link out to it.